megamek/src/megamek/common/Entity.java

```diff
@@ -9042,21 +9042,36 @@ public short getUnitNumber() {
         return unitNumber;
     }
 
+
+    /**
+     * Returns whether an entity is on the edge of the board.
+     */
+    public boolean atEdge() {
+    	Coords pos = getPosition();
+        return ((pos.getX() == 0) || (pos.getX() == (game.getBoard().getWidth() - 1))
+                || (pos.getY() == 0) || (pos.getY() == (game.getBoard()
+                        .getHeight() - 1)));
+    }
+    
+
+    /**
+     * Returns whether an entity currently has MP.
+     */
+    public boolean hasMP() {
+    	return (getWalkMP() > 0);
+    }
+    
     /**
-     * Returns whether an entity can flee from its current position. Currently
-     * returns true if the entity is on the edge of the board.
+     * Returns whether an entity can flee.
      */
     public boolean canFlee() {
-        Coords pos = getPosition();
-        return (pos != null)
-               && ((getWalkMP() > 0) || (this instanceof Infantry))
+    	Coords pos = getPosition();
```

Comment:
As per neoancient's comment, the default base class implementation should be canFlee() && hasMP() && atEdge(). That way, you don't have to override it for mechs, tanks, etc, just aeros and infantry.

Reply:
Both Infantry and Aero behave differently, and they're both subtractive
cases.  Originally I just added the Aero check to the base class evaluation by putting an or clause on the position that said if it was an aero at altitude 10 ignore the edge requirement.  People on slack thought it was better not to clutter Entity and instead put that stuff into the type specific classes, so in order to avoid duplicating the base code in aero and infantry I needed to break it out into components.  I don't see another way to take out the edge and MP requirements for those two cases..

Reply:
You can selectively invoke the helper functions you wrote in the base class from the inheriting classes, since you're overriding the base class implementation of canFlee(). The aero logic is purely additive, while, with the infantry, you can omit the call to "hasMP".

Reply:
ah so I add a new helper function that returns the current base class implementation of canFlee, called, say 'underControl' and then make the base class implementation of canFlee underControl and atEdge and hasMP, then override that function in aero as underControl, hasMP and altitude ==10, and the infantry implementation as underControl and atEdge?